libraries/AP_Networking/AP_Networking_CAN.cpp

```diff
@@ -0,0 +1,226 @@
+/*
+  CAN UDP multicast server
+ */
+
+#include "AP_Networking_Config.h"
+
+#if AP_NETWORKING_CAN_MCAST_ENABLED
+
+#include "AP_Networking.h"
+
+#include <AP_HAL/utility/Socket.h>
+#include <AP_HAL/CANIface.h>
+#include <GCS_MAVLink/GCS.h>
+#include <AP_Math/crc.h>
+
+#if CONFIG_HAL_BOARD == HAL_BOARD_CHIBIOS
+#include "hal.h"
+#include <AP_HAL_ChibiOS/hwdef/common/stm32_util.h>
+#include <AP_HAL_ChibiOS/CANIface.h>
+#endif
+
+#define MCAST_ADDRESS_BASE "239.65.82.0"
+#define MCAST_PORT 57732U
+#define MCAST_MAGIC 0x2934U
+#define MCAST_FLAG_CANFD 0x0001
+#define MCAST_MAX_PKT_LEN 74 // 64 byte data + 10 byte header
+
+struct PACKED mcast_pkt {
+    uint16_t magic;
+    uint16_t crc;
+    uint16_t flags;
+    uint32_t message_id;
+    uint8_t data[MCAST_MAX_PKT_LEN-10];
+};
+
+#define MCAST_HDR_LENGTH offsetof(mcast_pkt, data)
+
+extern const AP_HAL::HAL& hal;
+
+#ifdef HAL_BOOTLOADER_BUILD
+void AP_Networking_CAN::mcast_trampoline(void *ctx)
+{
+    auto *mcast = (AP_Networking_CAN *)ctx;
+    mcast->mcast_server();
+}
+extern ChibiOS::CANIface can_iface[HAL_NUM_CAN_IFACES];
+extern void thread_sleep_us(uint32_t us);
+#endif // HAL_BOOTLOADER_BUILD
+
+/*
+  get CAN interface for a bus
+ */
+AP_HAL::CANIface *AP_Networking_CAN::get_caniface(uint8_t bus) const
+{
+#ifdef HAL_BOOTLOADER_BUILD
+    return &can_iface[bus];
+#else
+    return hal.can[bus];
+#endif
+}
+
+/*
+  start the CAN multicast server
+ */
+void AP_Networking_CAN::start(uint8_t _bus_mask)
+{
+    const uint32_t stack_size = 8192;
+    bus_mask = _bus_mask;
+
+#ifdef HAL_BOOTLOADER_BUILD
+    thread_create_alloc(THD_WORKING_AREA_SIZE(stack_size),
+                        "CAN_MCAST",
+                        60,
+                        mcast_trampoline,
+                        this);
+#else
+    hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&AP_Networking_CAN::mcast_server, void),
+                                 "CAN_MCAST",
+                                 stack_size, AP_HAL::Scheduler::PRIORITY_CAN, -1);
+#endif
+}
+
+/*
+  main thread for CAN multicast server
+ */
+void AP_Networking_CAN::mcast_server(void)
+{
+#ifndef HAL_BOOTLOADER_BUILD
+    while (!hal.scheduler->is_system_initialized()) {
+        hal.scheduler->delay(100);
+    }
+#endif
+    GCS_SEND_TEXT(MAV_SEVERITY_INFO, "CAN_MCAST: starting");
+
+    ObjectBuffer<AP_HAL::CANFrame> *frame_buffers[HAL_NUM_CAN_IFACES] {};
+
+    for (uint8_t bus=0; bus<HAL_NUM_CAN_IFACES; bus++) {
+        auto *cbus = get_caniface(bus);
+        if (cbus == nullptr) {
+            continue;
+        }
+        if (bus_mask & (1U<<bus)) {
+            mcast_sockets[bus] = NEW_NOTHROW SocketAPM(true);
+            if (mcast_sockets[bus] == nullptr) {
+                GCS_SEND_TEXT(MAV_SEVERITY_ERROR, "CAN_MCAST[%u]: failed to create socket", unsigned(bus));
```

Comment:
Would  factoring out `CAN_MCAST[%u]:` into a single place save flash? I've seen this done in other places.

Reply:
i doubt it will save flash, and it would make it harder to read

Reply:
Ok. I'll go check on the other places I did this. It was harder to read, and never checked it actually saved flash.